toolchain/check/impl.cpp

```diff
@@ -227,17 +232,197 @@ auto ImplWitnessForDeclaration(Context& context, const SemIR::Impl& impl)
        .specific_id = context.generics().GetSelfSpecific(impl.generic_id)});
 }
 
-auto ImplWitnessStartDefinition(Context& /*context*/, SemIR::Impl& impl)
+// Returns `true` if the `FacetAccessWitness` of `witness_id` matches
+// `interface_id`.
+static auto WitnessAccessMatchesInterface(Context& context,
+                                          SemIR::InstId witness_id,
+                                          SemIR::InterfaceId interface_id)
+    -> bool {
+  auto access = context.insts().GetAs<SemIR::FacetAccessWitness>(witness_id);
+  auto type_id = context.insts().Get(access.facet_value_inst_id).type_id();
+  auto facet_type = context.types().GetAs<SemIR::FacetType>(type_id);
+  auto facet_info = context.facet_types().Get(facet_type.facet_type_id);
```

Comment:
This makes a copy of the `FacetTypeInfo`.
```suggestion
  const auto& facet_info = context.facet_types().Get(facet_type.facet_type_id);
```